### PR TITLE
IG-14794: handle 'nil' values

### PR DESCRIPTION
**Pull request by imakhlin**
~~~
deps:
   frames:
       git_url: git@github.com:imakhlin/frames.git
       branch: IG-14794
~~~

handle 'nil' values + add integration test; note this fix covers scope of single frame only

### CODE REVIEW

File: pkg/pquerier/frames.go [hunk 1]
```diff
@@ -511,20 +515,25 @@ func (d *dataFrame) rawSeriesToColumns() error {
...
 
 	for nonExhaustedIterators > 0 {
+		rowIndex := 0
 		currentTime = nextTime
 		nextTime = int64(math.MaxInt64)
 		timeData = append(timeData, time.Unix(currentTime/1000, (currentTime%1000)*1e6))
 
+		// add new row to null values map
+		d.nullValuesMaps = append(d.nullValuesMaps, &pb.NullValuesMap{NullColumns: make(map[string]bool)})
+		nullValuesRowIndex := len(d.nullValuesMaps) - 1
```
Comment (talIguaz) [inline pkg/pquerier/frames.go]: Isn't that eventually just `nullValuesRowIndex++` ?

File: pkg/pquerier/frames.go [hunk 2]
```diff
@@ -540,7 +549,9 @@ func (d *dataFrame) rawSeriesToColumns() error {
 				t, v = iter.At()
 			}
 
+			colName := columns[seriesIndex].Name()
 			if t == currentTime {
+				//d.nullValuesMaps[nullValuesRowIndex].NullColumns[colName] = false
```
Comment (talIguaz) [inline pkg/pquerier/frames.go]: delete

File: pkg/pquerier/frames.go [hunk 3]
```diff
@@ -540,7 +549,9 @@ func (d *dataFrame) rawSeriesToColumns() error {
 				t, v = iter.At()
 			}
 
+			colName := columns[seriesIndex].Name()
```
Comment (talIguaz) [inline pkg/pquerier/frames.go]: no need for the extra variable, you can just assign it in line 564

File: pkg/pquerier/frames.go [hunk 4]
```diff
@@ -511,20 +515,25 @@ func (d *dataFrame) rawSeriesToColumns() error {
...
 			columns[i] = frames.NewSliceColumnBuilder(rawSeries.Labels().Get(config.PrometheusMetricNameAttribute),
 				frames.FloatType, 0)
 			seriesToDataType[i] = frames.FloatType
-			seriesTodefaultValue[i] = math.NaN()
+			seriesToDefaultValue[i] = math.NaN()
 		}
 	}
 
 	for nonExhaustedIterators > 0 {
+		rowIndex := 0
```
Comment (talIguaz) [inline pkg/pquerier/frames.go]: what is that for?

File: pkg/pquerier/pqueriertest/dataframe_query_integration_test.go [hunk 5]
```diff
@@ -856,3 +856,106 @@ func (suite *testSelectDataframeSuite) TestQueryNonExistingMetric() {
...
+	labelsWithNameLog := append(labels, utils.LabelsFromStringList("__name__", metricLog)...)
+
+	expectedTimeColumn := []int64{
+		suite.basicQueryTime,
+		suite.basicQueryTime + tsdbtest.MinuteInMillis,
+		suite.basicQueryTime + 2*tsdbtest.MinuteInMillis,
+		suite.basicQueryTime + 3*tsdbtest.MinuteInMillis,
+		suite.basicQueryTime + 4*tsdbtest.MinuteInMillis}
+
+	expectedTimeColumnLog := []int64{
```
Comment (talIguaz) [inline pkg/pquerier/pqueriertest/dataframe_query_integration_test.go]: It's not expected. it's just the times you ingest data

File: pkg/pquerier/pqueriertest/dataframe_query_integration_test.go [hunk 6]
```diff
@@ -856,3 +856,106 @@ func (suite *testSelectDataframeSuite) TestQueryNonExistingMetric() {
 		}
 	}
 }
+
+func (suite *testSelectDataframeSuite) TestSparseStringAndNumericColumnsDataframe() {
```
Comment (talIguaz) [inline pkg/pquerier/pqueriertest/dataframe_query_integration_test.go]: Please also add tests to other cases, such as
1. where you have 2 labels and one metric is not present in one of the labels
2. where you request for a not existing metric
3. if you can think of something else..

File: pkg/pquerier/pqueriertest/dataframe_query_integration_test.go (hunk 5 above)
Comment (imakhlin) [reply pkg/pquerier/pqueriertest/dataframe_query_integration_test.go]: ok

File: pkg/pquerier/pqueriertest/dataframe_query_integration_test.go [hunk 7]
```diff
@@ -959,3 +959,353 @@ func (suite *testSelectDataframeSuite) TestSparseStringAndNumericColumnsDatafram
...
+func (suite *testSelectDataframeSuite) TestSparseNumericColumnsWithEmptyColumnsDataframe() {
+	labelSetLinux := utils.LabelsFromStringList("os", "linux")
+	labelSetWindows := utils.LabelsFromStringList("os", "windows")
+	expectedTimeColumn := []int64{
+		suite.basicQueryTime,
+		suite.basicQueryTime + tsdbtest.MinuteInMillis,
+		suite.basicQueryTime + 2*tsdbtest.MinuteInMillis,
+		suite.basicQueryTime + 3*tsdbtest.MinuteInMillis,
+		suite.basicQueryTime + 4*tsdbtest.MinuteInMillis}
+	expectedColumns := map[string][]interface{}{
```
Comment (talIguaz) [inline pkg/pquerier/pqueriertest/dataframe_query_integration_test.go]: I would use a composite key of `metric-label` something like `cpu_0-linux`, `cpu_1-linux`, `cpu_1-win`...
it's a lot clearer then comparing two DFs with 5 rows each against a 10 items list..

Comment (imakhlin) [reply pkg/pquerier/pqueriertest/dataframe_query_integration_test.go]: you're welcome to refactor this in separate PR

**Pull request merged**